### ISSUE TITLE
[filestorage] receiver name sanitization

### ISSUE THREAD

**Issue by newly12** [open-telemetry/opentelemetry-collector-contrib#20731]
### Is your feature request related to a problem? Please describe.

`/` might be used in receiver name, for example, `filelog/pod/container/stdout`, when storage is enabled(for persisting file fingerprints, offsets, etc), the path will be `receiver_filelog_pod/container/stdout`, and dir `receiver_filelog_pod/container` will need to be created first, I think it makes sense to avoid such issue via the sanitization instead of changing receiver name.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2e2623c4ce6198ceb7db9ecdd8bb82f535a07443/extension/storage/filestorage/extension.go#L58-L64

### Describe the solution you'd like

replace `/` with `~` in receiver name, looks like such convention is used in some k8s components, for example

* `pods/volumes/kubernetes.io~csi`
* `pods/volumes/kubernetes.io~empty~dir`
* `pods/volumes/kubernetes.io~secrets`

**Comment by astencel-sumo** [comment]
I can confirm this is an issue, here's a sample config:

```yaml
exporters:
  logging:

extensions:
  file_storage:
    directory: .

receivers:
  filelog/logs/json:
    include: ./logs-json.log
    start_at: beginning
    storage: file_storage

service:
  extensions:
  - file_storage
  pipelines:
    logs:
      exporters:
      - logging
      receivers:
      - filelog/logs/json
```

and the output for the current latest v0.75.0

```console
$ ./otelcol-contrib-0.75.0-darwin_amd64 --config ./config.yaml
2023-04-06T17:03:18.713+0200    info    service/telemetry.go:90 Setting up own telemetry...
2023-04-06T17:03:18.713+0200    info    service/telemetry.go:116        Serving Prometheus metrics      {"address": ":8888", "level": "Basic"}
2023-04-06T17:03:18.713+0200    info    exporter@v0.75.0/exporter.go:286        Development component. May change in the future.        {"kind": "exporter", "data_type": "logs", "name": "logging"}
2023-04-06T17:03:18.715+0200    info    service/service.go:129  Starting otelcol-contrib...     {"Version": "0.75.0", "NumCPU": 16}
2023-04-06T17:03:18.715+0200    info    extensions/extensions.go:41     Starting extensions...
2023-04-06T17:03:18.715+0200    info    extensions/extensions.go:44     Extension is starting...        {"kind": "extension", "name": "file_storage"}
2023-04-06T17:03:18.715+0200    info    extensions/extensions.go:48     Extension started.      {"kind": "extension", "name": "file_storage"}
2023-04-06T17:03:18.715+0200    info    adapter/receiver.go:56  Starting stanza receiver        {"kind": "receiver", "name": "filelog/logs/json", "data_type": "logs"}
2023-04-06T17:03:18.715+0200    info    service/service.go:155  Starting shutdown...
2023-04-06T17:03:18.715+0200    info    adapter/receiver.go:150 Stopping stanza receiver        {"kind": "receiver", "name": "filelog/logs/json", "data_type": "logs"}
2023-04-06T17:03:18.715+0200    info    extensions/extensions.go:55     Stopping extensions...
2023-04-06T17:03:18.715+0200    info    service/service.go:169  Shutdown complete.
Error: cannot start pipelines: storage client: open receiver_filelog_logs/json: no such file or directory
2023/04/06 17:03:18 collector server run finished with error: cannot start pipelines: storage client: open receiver_filelog_logs/json: no such file or directory
```

I agree this should be fixed by sanitizing the receiver's name by replacing the slashes with another character. How about the underscore `_`? It is already being used as the separator between the receiver type `filelog` and the instnce name.

**Comment by newly12** [comment]
Thanks for looking into this. I don't have any concern to use underscore.

**Comment by djaglowski** [comment]
I think we should use a character that is less likely to be one chosen by a user, because for example we would still have a collision between `a/b_c` and `a_b/c`. @newly12 suggested `~` which seems like a better choice to me.

**Comment by djaglowski** [comment]
Related issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3148

**Comment by astencel-sumo** [comment]
@djaglowski would you be OK with a PR that just replaces slashes `/` with tildes `~` or do you think we should take a broader approach?

**Comment by djaglowski** [comment]
@astencel-sumo, that would be an improvement on what we have now, so seems fine. I think we should leave the issue open though, until we're confident that the solution is robust.

**Comment by astencel-sumo** [comment]
Here's the PR:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20896

I propose to close this issue with this PR as it is a (sub)duplicate of #3148 that's only about slashes, but keep #3148 open. What do you think @djaglowski?

**Comment by astencel-sumo** [comment]
I have reworked the PR mentioned above to fix the more general issue #3148.
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20896
